### PR TITLE
Add support for GIMP and JASC palettes (with per-color alpha)

### PR DESCRIPTION
**Pull request by reaperrr**
Supersedes #13913.

Compared to #13913, following changes and additions were made:
- merged GIMP/JASC support into a single trait
- replaced the limited-use `Tileset` with `Tilesets` + `ExcludeTilesets`
- Added support for per-color alpha
- updated testcase commits to account for per-color alpha (RA effects, TS units) and make testing quicker

### CODE REVIEW

File: mods/cnc/rules/palettes.yaml [hunk 1]
```diff
@@ -23,6 +23,10 @@
 		Filename: temperat.pal
 		ShadowIndex: 4
 		AllowModifiers: false
+	PaletteFromGimpFile:
```
Comment (Phrohdoh) [inline mods/cnc/rules/palettes.yaml]: This needs to be `PaletteFromGimpOrJascFile`, right?

**Comment by Phrohdoh** [comment]
There are quite a few `PaletteFromGimpFile`s in yaml that need to be `PaletteFromGimpOrJascFile`.

File: mods/cnc/rules/palettes.yaml (hunk 1 above)
Comment (reaperrr) [reply mods/cnc/rules/palettes.yaml]: Stupid me, of course.

**Comment by reaperrr** [comment]
Updated.

**Review by GraionDilach — APPROVED** [review]
Cooooool. :+1:

**Comment by pchote** [comment]
Do you want to merge the example commit, or is that just a testcase? (IMO they belong in https://github.com/OpenRA/ArtSrc, not here)

File: OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+		[Desc("Don't load palette for these tileset IDs.")]
+		public readonly HashSet<string> ExcludeTilesets = new HashSet<string>();
+
+		[FieldLoader.Require]
+		[Desc("Name of the file to load.")]
+		public readonly string Filename = null;
+
+		[Desc("Map listed indices to shadow. Ignores previous color.")]
+		public readonly int[] ShadowIndex = { };
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: Modders can define shadows (including alpha) directly in the palette, so this is redundant.

File: OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+						if (!byte.TryParse(rgba[2], out b))
+							throw new InvalidDataException("Invalid B value: {0}".F(rgba[2]));
+
+						if (rgba.Length == 3 || (rgba.Length > 3 && !byte.TryParse(rgba[3], out a)))
+							a = 0;
+
+						if (r == 0 && g == 0 && b == 0)
+							colors[i] = 0;
+						else if (a == 0)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: Doesn't this prevent fully transparent colours?

Comment (pchote) [reply OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: The issues here could be solved by treating `r == 0 && g == 0 && b == 0 && a == 0` (transparent) as distinct from `r == 0 && g == 0 && b == 0` (black)

Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: But alpha of 0 means zero transparency...?

Comment (pchote) [reply OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: Sorry yes, I meant non-transparent black on the case above.

**Comment by reaperrr** [comment]
> Do you want to merge the example commit?

No, I'll rename the commit titles to make that more clear.

**Comment by reaperrr** [comment]
Updated.

Palette index 0 is now always converted to fully transparent regardless of original color, in favor of otherwise keeping the code as simple and index-hack-free as possible.
Removed ShadowIndex in favor of using a black or dark-grey color + alpha.

Also updated testcases:
- GIMP testcase is RA, with explosion effects showcasing per-color alpha (transparency increases the darker the explosion color gets).
- JASC testcase is TS, showcasing non-back index 0 being converted to transparent, black + 128 alpha serving as shadow (color index 1).

**Comment by reaperrr** [comment]
Rebased to fix the AppVeyor issue.

**Comment by reaperrr** [comment]
@GraionDilach is your +1 still valid after the updates?

**Comment by GraionDilach** [comment]
I'm not happy with the slot 0 aspect, Photoshop throws the unused color slots to the end of it's palette when saving an indexed image, which might throw people off.

**Comment by GraionDilach** [comment]
On a second thought, I'm fine with an explicit mention of the slot 0 hardcode in the trait description.

**Comment by reaperrr** [comment]
OK. We can always update the trait if it turns out to be an issue.

**Review by GraionDilach — APPROVED** [review]
:+1:

**Comment by Mailaender** [comment]
Cherry-picked to bleed without the test cases https://github.com/OpenRA/OpenRA/commit/12054506e1399bccadd94727299266fc3133821f.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/29d4d2961f33be061eafde10e518ad4f57789a80...bcf8c022dacc0208dad4bd02db211f9bbbd42da6)

**Pull request closed without merging**